src/tracer/MOM_offline_main.F90

```diff
@@ -598,92 +574,97 @@ subroutine offline_redistribute_residual(CS, h_pre, uhtr, vhtr, converged)
+!> Returns the sums of any non-negligible remaining transport [H L2 ~> m3 or kg] to check for advection convergence
+real function remaining_transport_sum(G, GV, US, uhtr, vhtr, h_new)
+  type(ocean_grid_type),      intent(in)    :: G     !< Ocean grid structure
+  type(verticalGrid_type),    intent(in)    :: GV    !< Vertical grid structure
+  type(unit_scale_type),      intent(in)    :: US    !< A dimensional unit scaling type
+  real, dimension(SZIB_(G),SZJ_(G),SZK_(GV)), &
+                              intent(in   ) :: uhtr  !< Zonal mass transport [H L2 ~> m3 or kg]
+  real, dimension(SZI_(G),SZJB_(G),SZK_(GV)), &
+                              intent(in   ) :: vhtr  !< Meridional mass transport [H L2 ~> m3 or kg]
+  real, dimension(SZI_(G),SZJB_(G),SZK_(GV)), &
+                              intent(in   ) :: h_new !< Layer thicknesses [H ~> m or kg m-2]
 
   ! Local variables
-  integer :: i, j, k
-  integer :: is, ie, js, je, nz
-  real :: h_min !< A layer thickness below roundoff from GV type
-  real :: uh_neglect !< A small value of zonal transport that effectively is below roundoff error
-  real :: vh_neglect !< A small value of meridional transport that effectively is below roundoff error
+  real, dimension(SZI_(G),SZJ_(G)) :: trans_rem_col !< The vertical sum of the absolute value of
+                     !! transports through the faces of a column, in MKS units [kg].
+  real :: trans_cell !< The sum of the absolute value of the remaining transports through the faces
+                     !! of a tracer cell [H L2 ~> m3 or kg]
+  real :: HL2_to_kg_scale !< Unit conversion factor to cell mass [kg H-1 L-2 ~> kg m-3 or 1]
+  integer :: i, j, k, is, ie, js, je, nz
 
-  nz = CS%GV%ke
-  is = CS%G%isc ; ie = CS%G%iec ; js = CS%G%jsc ; je = CS%G%jec
+  is = G%isc ; ie = G%iec ; js = G%jsc ; je = G%jec ; nz = GV%ke
 
-  h_min = CS%GV%H_subroundoff
+  HL2_to_kg_scale = GV%H_to_kg_m2 * US%L_to_m**2
 
-  remaining_transport_sum = 0.
+  trans_rem_col(:,:) = 0.0
   do k=1,nz ; do j=js,je ; do i=is,ie
-    uh_neglect = h_min*CS%G%US%L_to_m**2*MIN(CS%G%areaT(i,j),CS%G%areaT(i+1,j))
-    vh_neglect = h_min*CS%G%US%L_to_m**2*MIN(CS%G%areaT(i,j),CS%G%areaT(i,j+1))
-    if (ABS(uhtr(I,j,k))>uh_neglect) then
-      remaining_transport_sum = remaining_transport_sum + ABS(uhtr(I,j,k))
-    endif
-    if (ABS(vhtr(i,J,k))>vh_neglect) then
-      remaining_transport_sum = remaining_transport_sum + ABS(vhtr(i,J,k))
-    endif
+    trans_cell = (ABS(uhtr(I-1,j,k)) + ABS(uhtr(I,j,k))) + &
+                 (ABS(vhtr(i,J-1,k)) + ABS(vhtr(i,J,k)))
+    if (trans_cell > max(1.0e-16*h_new(i,j,k), GV%H_subroundoff) * G%areaT(i,j)) &
+      trans_rem_col(i,j) =  trans_rem_col(i,j) + HL2_to_kg_scale * trans_cell
   enddo ; enddo ; enddo
-  call sum_across_PEs(remaining_transport_sum)
+
+  ! The factor of 0.5 here is to avoid double-counting because two cells share a face.
+  remaining_transport_sum = 0.5 * GV%kg_m2_to_H*US%m_to_L**2 * &
```

Comment:
Nice catch

Reply:
Thank you, but the original code did not have this issue of double counting because it was working on the cell faces.  However, shifting the calculation to the cells made it much easier to ensure that consistent results are obtained from the global sum in either symmetric or non-symmetric memory mode.